### PR TITLE
add aws xray id generator

### PR DESCRIPTION
**Pull request by sreeo**
Fixes #
Design discussion issue (if applicable) #

## Changes
https://github.com/open-telemetry/opentelemetry-rust/issues/1489

Add Xray IdGenerator from opentelemetry-rust crate

## Merge requirement checklist

* [x] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-rust/blob/main/CONTRIBUTING.md) guidelines followed
* [x] Unit tests added/updated (if applicable)
* [x] Appropriate `CHANGELOG.md` files updated for non-trivial, user-facing changes
* [ ] Changes in public API reviewed (if applicable)

### CODE REVIEW

**Comment by cijothomas** [comment]
Can you add a changelog entry for this? Please state explicitly that this is ported over from the main package..

**Comment by sreeo** [comment]
> Can you add a changelog entry for this? Please state explicitly that this is ported over from the main package..

done

File: opentelemetry-contrib/CHANGELOG.md [hunk 1]
```diff
@@ -8,6 +8,7 @@
 
 - Bump MSRV to 1.65 [#1318](https://github.com/open-telemetry/opentelemetry-rust/pull/1318)
 - Bump MSRV to 1.64 [#1203](https://github.com/open-telemetry/opentelemetry-rust/pull/1203)
+- Move Xray IdGenerator from `opentelemetry-rust` to `opentelemetry-contrib`  [#1489]
```
Comment (TommyCpp) [inline opentelemetry-contrib/CHANGELOG.md:11]:
```suggestion
- Move Xray IdGenerator from `opentelemetry-rust` to `opentelemetry-contrib`  [#33](https://github.com/open-telemetry/opentelemetry-rust-contrib/pull/33)
```

File: opentelemetry-contrib/src/trace/id_generator/aws.rs [hunk 2]
```diff
@@ -0,0 +1,94 @@
+use opentelemetry::trace::{SpanId, TraceId};
+use opentelemetry_sdk::trace::{IdGenerator, RandomIdGenerator};
+use std::time::{Duration, UNIX_EPOCH};
+
+/// Generates AWS X-Ray compliant Trace and Span ids.
```
Comment (TommyCpp) [inline opentelemetry-contrib/src/trace/id_generator/aws.rs:5]: I suppose this part is copy paste?

Comment (sreeo) [reply opentelemetry-contrib/src/trace/id_generator/aws.rs:5]: yeah copied as is from the opentelemetry core repo.

Comment (sreeo) [reply opentelemetry-contrib/src/trace/id_generator/aws.rs:5]: some outdated references here (imports), i'll fix it.

**Comment by lalitb** [comment]
Just wondering if this should be part of opentelemetry-aws crate, or contrib crate - question to @open-telemetry/rust-maintainers ? Logically both AWS XRay Propagator and XRay IdGenerator can be within the same crate.

**Comment by cijothomas** [comment]
> Just wondering if this should be part of opentelemetry-aws crate, or contrib crate - question to @open-telemetry/rust-maintainers ? Logically both AWS XRay Propagator and XRay IdGenerator can be within the same crate.

opentelemetry-aws seems the logical option, but that crate is also unofficial. https://github.com/open-telemetry/opentelemetry-rust-contrib/blob/main/opentelemetry-aws/src/lib.rs#L1
We need to see if anyone would like to own and maintain it, otherwise remove it.

**Comment by sreeo** [comment]
> We need to see if anyone would like to own and maintain it, otherwise remove it.
is this restricted to the maintainers or do you accept volunteers like me as well? @cijothomas

**Comment by cijothomas** [comment]
> > We need to see if anyone would like to own and maintain it, otherwise remove it.
> 
> is this restricted to the maintainers or do you accept volunteers like me as well? @cijothomas

It is open to all, but there could be some requirements (which we haven't written down here yet). For example, one requirement is that you join the OpenTelemetry organization, and that you can respond to issues reported to the component in a reasonable time etc. It may take a while before we formally write these down, but until then we can continue with this crate.

**Comment by sreeo** [comment]
okay

**Comment by cijothomas** [comment]
> Just wondering if this should be part of opentelemetry-aws crate, or contrib crate - question to @open-telemetry/rust-maintainers ? Logically both AWS XRay Propagator and XRay IdGenerator can be within the same crate.

@sreeo I think this is better option, than putting it into contrib crate and doing another move later (Its unclear what the future of contrib crate is. It was created when the contrib repo was not available, so may not be required). Could you move this into the aws crate itself in this PR?

**Comment by sreeo** [comment]
sure, i'll move it to the aws crate.

**Comment by hdost** [comment]
> > > We need to see if anyone would like to own and maintain it, otherwise remove it.
> 
> > 
> 
> > is this restricted to the maintainers or do you accept volunteers like me as well? @cijothomas
> 
> 
> 
> It is open to all, but there could be some requirements (which we haven't written down here yet). For example, one requirement is that you join the OpenTelemetry organization, and that you can respond to issues reported to the component in a reasonable time etc. It may take a while before we formally write these down, but until then we can continue with this crate.

The general guidelines are here https://github.com/open-telemetry/community/blob/main/CONTRIBUTING.md this applies to the organization. 

Our SIGs contribution guidelines are https://github.com/open-telemetry/opentelemetry-rust-contrib/blob/main/CONTRIBUTING.md

However agreed we don't have a subset for these contrib crates. I think if we can have at least a set of approvers for the various contrib crates that would be great

**Review by hdost — CHANGES_REQUESTED** [review]
Thank you for the contribution! 

Small change needed.

File: opentelemetry-contrib/CHANGELOG.md [hunk 3]
```diff
@@ -12,6 +12,7 @@
 
 - Bump MSRV to 1.65 [#1318](https://github.com/open-telemetry/opentelemetry-rust/pull/1318)
 - Bump MSRV to 1.64 [#1203](https://github.com/open-telemetry/opentelemetry-rust/pull/1203)
+- Move Xray IdGenerator from `opentelemetry-rust` to `opentelemetry-contrib`  [#33](https://github.com/open-telemetry/opentelemetry-rust-contrib/pull/33)
```
[the remark this answers is not in the document]
Comment (sreeo) [reply opentelemetry-contrib/CHANGELOG.md]: fixed.

**Comment by lalitb** [comment]
Can you also update the [Readme.md](https://github.com/open-telemetry/opentelemetry-rust-contrib/blob/main/opentelemetry-aws/README.md#supported-component) for this opentelemetry-aws crate to indicate it contains both XRay propagator and XRay Id Generator. As of now, it says:

### Supported component
Currently, this crate only supports XRay propagator. Contributions are welcome.

**Review by TommyCpp — APPROVED** [review]

**Comment by cijothomas** [comment]
> I think if we can have at least a set of approvers for the various contrib crates that would be great

This is the tricky part.. It is not sufficient for someone to be listed in codeowners file! They need to get write permission to the repo itself :( This is generally an issue in other otel repos too, with everyone doing own ways to manage it. We'll need to invent something of our own or follow some other languages. But this is not something we need to fix right away.

**Comment by cijothomas** [comment]
Thanks @sreeo for taking care of this!

File: opentelemetry-aws/CHANGELOG.md [hunk 4]
```diff
@@ -29,6 +29,7 @@
 
 - reduce `tokio` feature requirements #750
 - Update to opentelemetry v0.18.0
+- Move Xray IdGenerator from `opentelemetry-rust` to `opentelemetry-aws` [#33](https://github.com/open-telemetry/opentelemetry-rust-contrib/pull/33)
```
Comment (cijothomas) [inline opentelemetry-aws/CHANGELOG.md:32]: the entry got into the wrong place. Could you fix that in a follow up?

Comment (sreeo) [reply opentelemetry-aws/CHANGELOG.md:32]: created #38

**Pull request merged**